bundles/org.openhab.binding.somfymylink/src/main/java/org/openhab/binding/somfymylink/internal/handler/SomfyMyLinkBridgeHandler.java

```diff
@@ -0,0 +1,549 @@
+        }
+
+        try {
+            SomfyMyLinkCommandShadeUp cmd = new SomfyMyLinkCommandShadeUp(targetId, config.systemId);
+            sendCommand(cmd);
+        } catch (SomfyMyLinkException e) {
+            logger.info("Error commanding shade up: {}", e.getMessage());
+            throw new SomfyMyLinkException("Error commanding shade up", e);
+        }
+    }
+
+    public void commandShadeDown(String targetId) throws SomfyMyLinkException {
+        SomfyMyLinkConfiguration config = this.config;
+        if (config == null || StringUtils.isEmpty(config.systemId)) {
+            throw new SomfyMyLinkException("Config not setup correctly. System Id is not set.");
+        }
+
+        try {
+            SomfyMyLinkCommandShadeDown cmd = new SomfyMyLinkCommandShadeDown(targetId, config.systemId);
+            sendCommand(cmd);
+        } catch (SomfyMyLinkException e) {
+            logger.info("Error commanding shade down: {}", e.getMessage());
+            throw new SomfyMyLinkException("Error commanding shade down", e);
+        }
+    }
+
+    public void commandShadeStop(String targetId) throws SomfyMyLinkException {
+        SomfyMyLinkConfiguration config = this.config;
+        if (config == null || StringUtils.isEmpty(config.systemId)) {
+            throw new SomfyMyLinkException("Config not setup correctly. System Id is not set.");
+        }
+
+        try {
+            SomfyMyLinkCommandShadeStop cmd = new SomfyMyLinkCommandShadeStop(targetId, config.systemId);
+            sendCommand(cmd);
+        } catch (SomfyMyLinkException e) {
+            logger.info("Error commanding shade stop: {}", e.getMessage());
+            throw new SomfyMyLinkException("Error commanding shade stop", e);
+        }
+    }
+
+    public void commandScene(Integer sceneId) throws SomfyMyLinkException {
+        SomfyMyLinkConfiguration config = this.config;
+        if (config == null || StringUtils.isEmpty(config.systemId)) {
+            throw new SomfyMyLinkException("Config not setup correctly. System Id is not set.");
+        }
+        
+        try {
+            SomfyMyLinkCommandSceneSet cmd = new SomfyMyLinkCommandSceneSet(sceneId, config.systemId);
+            sendCommand(cmd);
+        } catch (SomfyMyLinkException e) {
+            logger.info("Error commanding scene run: {}", e.getMessage());
+            throw new SomfyMyLinkException("Error commanding scene run", e);
+        }
+    }
+
+    private CompletableFuture<@Nullable Void> sendCommand(SomfyMyLinkCommandBase command) {
+        String json = gson.toJson(command);
+        return sendCommand(json);
+    }
+
+    private CompletableFuture<@Nullable Void> sendCommand(String command) {
+        CompletableFuture<@Nullable Void> future = new CompletableFuture<>();
+        ExecutorService commandExecutor = this.commandExecutor;
+        if (commandExecutor != null) {
+            commandExecutor.execute(() -> {
+                try {
```

Comment:
These can be optimized a little bit:
```suggestion
    private CompletableFuture<@Nullable Void> sendCommand(SomfyMyLinkCommandBase commandBase) {
        CompletableFuture<@Nullable Void> future = new CompletableFuture<>();
        ExecutorService commandExecutor = this.commandExecutor;
        if (commandExecutor != null) {
            String command = gson.toJson(commandBase);
            commandExecutor.execute(() -> {
                try {
```

You could take this a bit further and have gson directly write out to the socket outputstream if you want.

Reply:
done